### PR TITLE
BZ1918270 - update Azure storage class kind annotations in docs

### PR DESCRIPTION
**Pull request by bobfuru**
[BZ1918270](https://bugzilla.redhat.com/show_bug.cgi?id=1918270) 
- Changed annotation to denote that `kind: Dedicated` in Azure Disk storage class is not supported by Red Hat. 
- Updated recommendation in Azure File to use `volumeBindingMode: WaitForFirstConsumer`.
- Removes some hard-wraps

For OCP 4.4+.

**PREVIEW LINKS:**

- Adds note that `kind: Dedicated` is not supported by Red Hat:
**Post-installation configuration -> Storage configuration -> Azure Disk object definition** https://deploy-preview-28932--osdocs.netlify.app/openshift-enterprise/latest/post_installation_configuration/storage-configuration.html#azure-disk-definition_post-install-storage-configuration

- Adds note that `kind: Dedicated` is not supported by Red Hat:
**Storage -> Persistent storage using Azure Disk -> Creating the Azure storage class** - https://deploy-preview-28932--osdocs.netlify.app/openshift-enterprise/latest/storage/persistent_storage/persistent-storage-azure.html#storage-create-azure-storage-class_persistent-storage-azure

- Updates doc to `volumeBindingMode: WaitForFirstConsumer`:
**Storage -> Dynamic provisioning -> Azure Disk object definition** - https://deploy-preview-28932--osdocs.netlify.app/openshift-enterprise/latest/storage/dynamic-provisioning.html#azure-file-definition_dynamic-provisioning

- Updates doc to `volumeBindingMode: WaitForFirstConsumer`:
**Storage -> Dynamic provisioning -> Azure Disk object definition** - https://deploy-preview-28932--osdocs.netlify.app/openshift-enterprise/latest/storage/dynamic-provisioning.html#azure-disk-definition_dynamic-provisioning

### CODE REVIEW

**Comment by bobfuru** [comment]
**PREVIEW LINKS:**

- Adds note that `kind: Dedicated` is not supported by Red Hat:
**Post-installation configuration -> Storage configuration -> Azure Disk object definition** https://deploy-preview-28932--osdocs.netlify.app/openshift-enterprise/latest/post_installation_configuration/storage-configuration.html#azure-disk-definition_post-install-storage-configuration

- Adds note that `kind: Dedicated` is not supported by Red Hat:
**Storage -> Persistent storage using Azure Disk -> Creating the Azure storage class** - https://deploy-preview-28932--osdocs.netlify.app/openshift-enterprise/latest/storage/persistent_storage/persistent-storage-azure.html#storage-create-azure-storage-class_persistent-storage-azure

- Updates doc to `volumeBindingMode: WaitForFirstConsumer`:
**Storage -> Dynamic provisioning -> Azure Disk object definition** - https://deploy-preview-28932--osdocs.netlify.app/openshift-enterprise/latest/storage/dynamic-provisioning.html#azure-file-definition_dynamic-provisioning

- Updates doc to `volumeBindingMode: WaitForFirstConsumer`:
**Storage -> Dynamic provisioning -> Azure Disk object definition** - https://deploy-preview-28932--osdocs.netlify.app/openshift-enterprise/latest/storage/dynamic-provisioning.html#azure-disk-definition_dynamic-provisioning

**Comment by bobfuru** [comment]
@jsafrane PTAL for SME review and LMK if the "unsupported" wording accurately captures your recommendation.

**Comment by bobfuru** [comment]
@qinpingli PTAL for QE review

File: modules/dynamic-provisioning-azure-disk-definition.adoc
Comment (bobfuru) [inline modules/dynamic-provisioning-azure-disk-definition.adoc:36]: Based on [this comment](https://bugzilla.redhat.com/show_bug.cgi?id=1918270#c7) in the BZ.

File: modules/dynamic-provisioning-azure-file-considerations.adoc
Comment (bobfuru) [inline modules/dynamic-provisioning-azure-file-considerations.adoc:37]: Based on [this comment](https://bugzilla.redhat.com/show_bug.cgi?id=1918270#c2) in the BZ.

File: modules/dynamic-provisioning-azure-file-definition.adoc
Comment (bobfuru) [inline modules/dynamic-provisioning-azure-file-definition.adoc:59]: Based on [this comment](https://bugzilla.redhat.com/show_bug.cgi?id=1918270#c2) in the BZ.

File: modules/storage-azure-create-storage-class.adoc
Comment (bobfuru) [inline modules/storage-azure-create-storage-class.adoc:40]: Based on [this comment](https://bugzilla.redhat.com/show_bug.cgi?id=1918270#c7) in the BZ.

File: modules/dynamic-provisioning-azure-disk-definition.adoc
Comment (qinpingli) [inline modules/dynamic-provisioning-azure-disk-definition.adoc:20]: @bobfuru  How about using storageclass defined in https://github.com/openshift/cluster-storage-operator/blob/master/assets/storageclasses/azure.yaml?

File: modules/dynamic-provisioning-azure-disk-definition.adoc
Comment (qinpingli) [inline modules/dynamic-provisioning-azure-disk-definition.adoc:22]: With the new storageclass, seems this is unnecessary.

File: modules/dynamic-provisioning-azure-disk-definition.adoc
Comment (qinpingli) [reply modules/dynamic-provisioning-azure-disk-definition.adoc:36]: @bobfuru From line 26, seems  the volumes created with `kind: Shared` are the unmanaged disks too.

File: modules/dynamic-provisioning-azure-file-considerations.adoc
Comment (qinpingli) [reply modules/dynamic-provisioning-azure-file-considerations.adoc:37]: @bobfuru IIUC, for azure file, we don't need to set volumeBindingMode to WaitForFirstConsumer, azure file shares should no the limit as azure disk in this comment. The storageclass I recommended uses the correct volumeBindingMode.

File: modules/dynamic-provisioning-azure-file-definition.adoc
Comment (qinpingli) [reply modules/dynamic-provisioning-azure-file-definition.adoc:59]: Same as the last comment.

File: modules/dynamic-provisioning-azure-file-definition.adoc
Comment (qinpingli) [inline modules/dynamic-provisioning-azure-file-definition.adoc:68]: Looks like it's unnecessary.

File: modules/storage-azure-create-storage-class.adoc
Comment (qinpingli) [reply modules/storage-azure-create-storage-class.adoc:40]: seems the volumes created with kind: Shared are the unmanaged disks too.

**Comment by qinpingli** [comment]
@bobfuru Thanks for the update, added some comments.

File: modules/dynamic-provisioning-azure-disk-definition.adoc
Comment (bobfuru) [reply modules/dynamic-provisioning-azure-disk-definition.adoc:20]: Love it. Done.

File: modules/dynamic-provisioning-azure-disk-definition.adoc
Comment (bobfuru) [reply modules/dynamic-provisioning-azure-disk-definition.adoc:22]: Removed

File: modules/dynamic-provisioning-azure-disk-definition.adoc
Comment (bobfuru) [reply modules/dynamic-provisioning-azure-disk-definition.adoc:36]: @jsafrane IIUC, @qinpingli is indicating here that RH would not support `kind: Shared` either because Azure creates unmanaged disks in that scenario as well. Is that correct, and if so, do we need to update this note to something like the following?

> Red Hat only supports the use of `kind: Managed` in the storage class...etc.

File: modules/dynamic-provisioning-azure-file-considerations.adoc
Comment (bobfuru) [reply modules/dynamic-provisioning-azure-file-considerations.adoc:37]: @qinpingli - I'm confused. The storage class you recommended above for Azure Disk also sets the `volumeBindingMode` to `WaitForFirstConsumer` (https://github.com/openshift/cluster-storage-operator/blob/1842b2a59771d1425272789c080619cad06a2f3c/assets/storageclasses/azure.yaml#L8). Are you saying that I should revert to using `Immediate` for this field?

File: modules/dynamic-provisioning-azure-file-definition.adoc
Comment (bobfuru) [reply modules/dynamic-provisioning-azure-file-definition.adoc:59]: Same question as previous, should this revert back to `Immediate`?

File: modules/dynamic-provisioning-azure-file-definition.adoc
Comment (bobfuru) [reply modules/dynamic-provisioning-azure-file-definition.adoc:68]: IIUC, the "is recommended" note should only be used for Azure Disk. So, I have removed it here and relocated to the Azure Disk definition example

File: modules/storage-azure-create-storage-class.adoc
Comment (bobfuru) [reply modules/storage-azure-create-storage-class.adoc:40]: Same question here for @jsafrane: Do we need to update this note to something like the following?

> Red Hat only supports the use of `kind: Managed` in the storage class...etc.

**Comment by bobfuru** [comment]
Thanks for the review, @qinpingli. I've addressed what I could but need further clarification on a few things from you and @jsafrane.

File: modules/dynamic-provisioning-azure-file-considerations.adoc
Comment (qinpingli) [reply modules/dynamic-provisioning-azure-file-considerations.adoc:37]: @bobfuru There is some detailed explanation: https://kubernetes.io/docs/concepts/storage/storage-classes/#volume-binding-mode.

Although from the simple test, looks like `WaitForFirstConsumer` binding mode is harmless for azure file, but this mode is not introduced for azure file.

Comment (jsafrane) [reply modules/dynamic-provisioning-azure-file-considerations.adoc:37]: Indeed, for Azure File, `WaitForFirstConsumer` is not better than `Immediate` . It's not worse either.

For AzureDisk we use `WaitForFirstConsumer`, because Azure Disk volumes can be used only by nodes in the same *zone* as the volume, therefore the volume should be provisioned in a zone that has free nodes.

For AzureFile it does not matter that much, because Azure File volumes can be used only by nodes in the same *zone* as the volume can be accessed from any zone and users don't need to worry about volume placement.

File: modules/dynamic-provisioning-azure-disk-definition.adoc
Comment (jsafrane) [reply modules/dynamic-provisioning-azure-disk-definition.adoc:40]: I haven't been able to create volumes with `kind: Shared`, but I don't know if it was mis-configuration on my / OCP side or it really can't be supported in OCP. Unfortunately, I did not save the error message.

@qinpingli do you know if we can support `Shared` volumes?

Comment (qinpingli) [reply modules/dynamic-provisioning-azure-disk-definition.adoc:40]: @jsafrane Checked our test cases, looks like no cases to set kind = `Shared` explicitly. 
From the [doc](https://kubernetes.io/docs/concepts/storage/storage-classes/#volume-binding-mode), looks like the default kind is `Shared`, tried with the following storageclass:
$ cat sc-default.yaml 
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
  name: slow
provisioner: kubernetes.io/azure-disk
parameters:
  skuName: Standard_LRS
  location: centralus
  storageAccount: XXXXX
Volume can be created successfully, and be attached successfully.

Tried the following storageclass to set the kind explicitly:
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
  name: slow
provisioner: kubernetes.io/azure-disk
parameters:
  skuName: Standard_LRS
  location: centralus
  storageAccount: XXXXX
  kind: Shared
Volume can not be created, get the following event from PVC:
Failed to provision volume with StorageClass "slow": couldn't find key named key1 in storage account:XXXX keys

seems the default kind is not `Shared`, and ocp doesn't support it?

File: modules/dynamic-provisioning-azure-file-considerations.adoc
Comment (bobfuru) [reply modules/dynamic-provisioning-azure-file-considerations.adoc:37]: Thanks both for the thorough explanation. I've reverted to `Immediate` so that this field isn't updated needlessly.

File: modules/dynamic-provisioning-azure-file-definition.adoc
Comment (bobfuru) [reply modules/dynamic-provisioning-azure-file-definition.adoc]: Also reverted to `Immediate` here to maintain consistency in this field for Azure file docs.

File: modules/dynamic-provisioning-azure-disk-definition.adoc
Comment (bobfuru) [reply modules/dynamic-provisioning-azure-disk-definition.adoc:40]: @sferich888 It looks like OCP cannot support `kind:Shared` or `kind:Dedicated` for Azure Disk, so I'm looking at adding a note to explain this. 

Do you have any concerns if we were to add a statement here along the lines of the following?

> **!**
> Red Hat only supports the use of `kind: Managed` in the storage class.
> 
> With `Shared` and `Dedicated`, Azure creates unmanaged disks, while {product-title} creates a managed disk for machine OS (root) disks. But because Azure Disk does not allow the use of both managed and unmanaged disks on a node, unmanaged disks created with `Shared` or `Dedicated` cannot be attached to {product-title} nodes.

File: modules/storage-azure-create-storage-class.adoc
Comment (bobfuru) [reply modules/storage-azure-create-storage-class.adoc:40]: Same question here for @sferich888 - looking for comments or ack to add a note to indicate that RH only supports `kind: Managed` and not `Shared` or `Dedicated`

File: modules/dynamic-provisioning-azure-disk-definition.adoc
Comment (jsafrane) [reply modules/dynamic-provisioning-azure-disk-definition.adoc]: I checked the upstream docs, 

> When kind is shared, all unmanaged disks are created in a few shared storage accounts in the same resource group as the cluster

This implies that shared disk is unmanaged too and can't work in OCP, with the same explanation as for `Dedicated` disks. The note in the previous comment looks good to me.

Comment (bobfuru) [reply modules/dynamic-provisioning-azure-disk-definition.adoc]: Thanks for confirming, @jsafrane. 

FYI, this will need to go through our change management process, requiring an ack from @sferich888 or @reestr prior to merge.

Comment (sferich888) [reply modules/dynamic-provisioning-azure-disk-definition.adoc]: @davemulford your probably best to review this due to its connection to Azure. 

However as I understand it - were simply clarifying that due to the way OpenShift (in IPI/UPI deployments deploys), customers on Azure are limited to a specific disk type for PV's (when using the Azure Dynamic Storage Provider (as a storage class)). 

From a docs point of view: https://docs.microsoft.com/en-us/azure/virtual-machines/managed-disks-overview and https://docs.microsoft.com/en-us/azure/virtual-machines/disks-types (may be keep partner docs to link to) to help customers understand what it is were talking about. 

If it were me I would ack this, but I think @davemulford should review too.

Comment (sferich888) [reply modules/dynamic-provisioning-azure-disk-definition.adoc]: @jsafrane why doen't the product (API's) or operators; enforce this ? IE; why don't we use validations to say; this isn't a legal config on OpenShift?

Comment (davemulford) [reply modules/dynamic-provisioning-azure-disk-definition.adoc]: The wording is correct and technically accurate, so the docs change is fine.

I do agree with @sferich888 that we should be validating this in some form.  Whether that's through the product itself, or an operator, doesn't really matter to me.

**Comment by bobfuru** [comment]
@openshift/team-documentation PTAL

File: modules/dynamic-provisioning-azure-disk-definition.adoc
Comment (jsafrane) [reply modules/dynamic-provisioning-azure-disk-definition.adoc]: @sferich888, traditionally, StorageClass.Parameters are not validated anywhere in Kubernetes and OpenShift. We could add a webhook / admission plugin, but IMO it could break some user automation scripts / operators - objects previously accepted by API server would now be rejected. Those probably deserve to be broken though.

In addition, it will become endless whack-a-mole when upstream introduces new parameters.

File: modules/dynamic-provisioning-azure-disk-definition.adoc
Comment (jeana-redhat) [inline modules/dynamic-provisioning-azure-disk-definition.adoc:36]: General note on callouts: for this third one, the item is defined ("Azure storage account SKU tier.") and then the default and other info is given. Should the first two also start with a brief mention of what they are?

**Review by jeana-redhat — APPROVED** [review]
lgtm

File: modules/dynamic-provisioning-azure-disk-definition.adoc
Comment (bobfuru) [reply modules/dynamic-provisioning-azure-disk-definition.adoc:36]: It's a valid question. I see a mix in general for callouts. In this case, I think it defines it because this callout is referring to the SKU tier (`Premium_LRS` or `Standard_LRS`). The others are calling out the recommended formatting or parameter types. In other words, there's not a standard? ¯\_(ツ)_/¯ 
I'll leave it as stated but definitely worth pondering!

File: modules/storage-azure-create-storage-class.adoc
Comment (jsafrane) [reply modules/storage-azure-create-storage-class.adoc]: Yes, it should say that Shared is not supported too.

**Pull request merged**
